### PR TITLE
UCP: adds ucp_rkey_ptr()

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe @shamisp  please review

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4196/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4199/ for details (Mellanox internal link).

File: src/ucp/core/ucp_rkey.c [hunk 1]
```diff
@@ -205,6 +205,28 @@ ucs_status_t ucp_ep_rkey_unpack(ucp_ep_h ep, void *rkey_buffer, ucp_rkey_h *rkey
...
+    num_rkeys = ucs_count_one_bits(rkey->md_map);
+
+    for (i = 0; i < num_rkeys; ++i) {
+        status = uct_rkey_ptr(&rkey->uct[i], raddr, addr_p);
+        if ((status == UCS_OK) ||
+            (status == UCS_ERR_INVALID_ADDR)) {
+            return status;
+        }
+    }
+    return UCS_ERR_UNSUPPORTED;
```
Comment (yosefe) [inline src/ucp/core/ucp_rkey.c]: maybe better return UCS_ERR_UNREACHABLE?
also pls add space line before

File: test/gtest/ucp/test_ucp_mmap.cc [hunk 2]
```diff
@@ -53,6 +54,12 @@ void test_ucp_mmap::test_rkey_management(entity *e, ucp_mem_h memh, bool is_dumm
     }
     ASSERT_UCS_OK(status);
 
+    status = ucp_rkey_ptr(rkey, (uint64_t)memh->address, &ptr);
+    if (status == UCS_OK) {
+        EXPECT_EQ(memcmp(memh->address, ptr, memh->length), 0);
+    } else {
+        EXPECT_EQ(UCS_ERR_UNSUPPORTED, status);
+    }
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_mmap.cc]: pls add space line after

File: test/gtest/ucp/test_ucp_mmap.cc [hunk 3]
```diff
@@ -53,6 +54,12 @@ void test_ucp_mmap::test_rkey_management(entity *e, ucp_mem_h memh, bool is_dumm
     }
     ASSERT_UCS_OK(status);
 
+    status = ucp_rkey_ptr(rkey, (uint64_t)memh->address, &ptr);
+    if (status == UCS_OK) {
+        EXPECT_EQ(memcmp(memh->address, ptr, memh->length), 0);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_mmap.cc]: should be `EXPECT_EQ(0, memcmp(memh->address, ptr, memh->length))`

**Comment by yosefe** [comment]
@shamisp @bbenton pls take a look

**Comment by bbenton** [comment]
:+1:  for Doc/API

**Comment by yosefe** [comment]
@shamisp can we merge?

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -1620,6 +1620,25 @@ ucs_status_t ucp_ep_rkey_unpack(ucp_ep_h ep, void *rkey_buffer, ucp_rkey_h *rkey
 
 /**
  * @ingroup UCP_MEM
+ * @brief Get a local pointer to remote memory.
+ *
+ * This routine returns a local pointer to the remote memory described
+ * by the rkey.
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: please add a @note that the function will not work for all endpoints but only for a class of endpoints that are reachable through shared memory

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4240/ for details (Mellanox internal link).

**Comment by alex-mikheev** [comment]
:bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4241/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:bgate:retest

**Comment by shamisp** [comment]
👍

**Comment by yosefe** [comment]
bot:bgate:retest

**Pull request merged**